### PR TITLE
 Fix and cover RMSD

### PR DESCRIPTION
**Pull request by jdetle**
Fixes: #897

Changes made in this Pull Request:
 - Fixed a bug where we are still calling `_process_selection` as opposed to `process_selection` in RMSD (one that I think I created a while back)
- Changed RMSD to save numpy array.
- Added rmsdArray testfile
- Added RMSD tests

# TODO
- [x] Cover exceptions in RMSD
- [x] Cover logger warnings
- [x] Revert to savetxt
- [x] Change to psf, dcd, start, stop, step to speed up tests
- [x] Figure out how to load txt files as np arrays to compare, or some other way to test save function
- [x]  Test when reference not none in rmsd
- [x] Assert selection error  when reference not equal in length  to traj atoms
- [x] Same for group selections
- [x] Test mass_weighted in `run`
- [x] Rename rmsd tests
- [x] Undo update of changelog

Currently left for another PR
-----
+ Cover process_selection errors
+ Figure out how to force an overflow underflow error in RMSF

Currently stuck
------
+ Can open file handles be closed, if so, how?
+  Should tests always assert something? For example, I have a test to cover saving and weighting, but I don't actually check if the result is equal to some expected value.

### CODE REVIEW

**Comment by kain88-de** [comment]
why wasn't this raising an error in the tests?  Isn't this covered in the test?

**Comment by jdetle** [comment]
This [coverage report](https://coveralls.io/builds/6894396/source?filename=miniconda%2Fenvs%2Fpyenv%2Flib%2Fpython2.7%2Fsite-packages%2FMDAnalysis%2Fanalysis%2Frms.py) seems to indicate not! . I will go ahead and write some tests for the RMSD class.

**Comment by jdetle** [comment]
It seems like RMSD is ripe for a refactor to the Bauhaus API too, so I could just go ahead and do that as well.

**Comment by kain88-de** [comment]
Yes if you like. Please finish the dmap first though

**Comment by orbeckst** [comment]
Leave refactor for later (as Max said). Fixing the bug and some tests are good enough for this PR.

Re: naming: I don't like RMSDSingleFrame, it still looks like a class. But ultimately your choice, just do it :-).

**Comment by jdetle** [comment]
On it.

**Comment by jdetle** [comment]
Now is as good as a time as any, what are open file handles and how do I close them?

File: package/CHANGELOG [hunk 1]
```diff
@@ -25,6 +25,7 @@ Fixes
     next (Issue #869)
   * Display of Deprecation warnings doesn't affect other modules anymore (Issue #754)
   * Changed nframes to n_frames in analysis modules for consistency (Issue #890)
+  * Removed call to protected process_selection function in RMSD
```
Comment (orbeckst) [inline package/CHANGELOG]: Leave this out from the CHANGELOG. I don't think that this was broeken in 0.15.0 or was it? If it was broken then this would be an entry under fixes, just stating that rms.RMSD was fixed (and reference #897)

Comment (jdetle) [inline package/CHANGELOG]: Good point! Thanks.

File: package/MDAnalysis/analysis/rms.py [hunk 2]
```diff
@@ -527,7 +527,7 @@ def save(self, filename=None):
         if filename is not None:
             if self.rmsd is None:
                 raise NoDataError("rmsd has not been calculated yet")
-            np.savetxt(filename, self.rmsd)
+            np.save(filename, self.rmsd)
```
Comment (orbeckst) [inline package/MDAnalysis/analysis/rms.py]: Why did you change this? This can break code for users, don't do this light-heartedly! Please revert!

save and savetxt are two very different things.

File: testsuite/MDAnalysisTests/analysis/test_rms.py [hunk 3]
```diff
@@ -128,6 +128,44 @@ def test_with_superposition_equal(self):
         rmsd_superposition = rms.rmsd(A, B, center=True, superposition=True)
         assert_almost_equal(rmsd, rmsd_superposition)
 
+class TestRMSD(object):
+    def setUp(self):
+        self.universe = MDAnalysis.Universe(GRO, XTC)
```
Comment (orbeckst) [inline testsuite/MDAnalysisTests/analysis/test_rms.py]: use PSF and DCD --- much faster for the tests!

Comment (orbeckst) [inline testsuite/MDAnalysisTests/analysis/test_rms.py]: `Universe(PSF, DCD)` loads much faster than the GRO file.

To cut down on processing, use `start`, `stop`, and `step`, with the additional advantage that they get tested, too.

**Comment by jdetle** [comment]
There are a few mistakes and additional tests that I will go ahead and work on tonight. Beyond that, packing for a week in the bay before ***SCIPY 2016***

**Comment by jdetle** [comment]
Spurious test failures

**Comment by orbeckst** [comment]
Rerun the test on Travis. There should be a little reload button in the upper right. At least there is one for me when I have to do this. 

--
Oliver Beckstein
email: orbeckst@gmail.com

**Comment by jdetle** [comment]
I have it but for some reason it throws an error saying the job can't be restarted, it's either a bug or a permissions issue.

**Comment by jdetle** [comment]
I will be away from the computer until late tonight. I hope everything looks alright. 1% coverage increase has to be nice :)

**Comment by orbeckst** [comment]
Travis appears to be happy now. Everything passed.

**Comment by orbeckst** [comment]
Still WIP or ready to be merged?

**Comment by jdetle** [comment]
Ready to be merged. All the todos are todone. There are still some minor
questions that I have about testing but merge-able nonetheless.
On Sat, Jul 9, 2016 at 3:01 PM Oliver Beckstein <notifications@github.com>
wrote:

> Still WIP or ready to be merged?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/MDAnalysis/mdanalysis/pull/889#issuecomment-231558207>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AKcARg7JHi8iW2fNW77Hlq6usB7RvXrVks5qUBo_gaJpZM4JGYJ2>
> .
>
-- 
Have a wonderful day,
John J. Detlefs

**Comment by orbeckst** [comment]
Ah, just read the top post where you said:

1. Can open file handles be closed, if so, how?
2. Should tests always assert something? For example, I have a test to cover saving and weighting, but I don't actually check if the result is equal to some expected value

Answers:
1. Good question, @jbarnoud 's plugin reports files that are still open but I also don't have a clue how to fix these reported open files.
2. Each test should always assert something. At a minimum, check that a file was written, but better (and that's what we do in the trajectory writer tests), read the file back in and compare to the data that you used to write the file. If that is not feasible, manually read the file, compute some value (eg a mean) and check that you get it back from the saved file in the test.

File: package/MDAnalysis/analysis/rms.py [hunk 4]
```diff
@@ -83,7 +83,7 @@
 
    import matplotlib.pyplot as plt
    rmsd = R.rmsd.T   # transpose makes it easier for plotting
-   time = rmsd[1]
+   time = rmsd[1]facebook.com
```
Comment (orbeckst) [inline package/MDAnalysis/analysis/rms.py]: There's a random "facebook.com" in the docs. Please remove.

File: package/MDAnalysis/analysis/rms.py [hunk 5]
```diff
@@ -351,7 +351,7 @@ def __init__(self, traj, reference=None, select='all',
         self.ref_atoms = self.reference.select_atoms(*self.select['reference'])
         self.traj_atoms = self.universe.select_atoms(*self.select['mobile'])
         if len(self.ref_atoms) != len(self.traj_atoms):
-            logger.exception()
+            logger.exception('SelectionError: ref_atoms, traj_atoms unequal')
```
Comment (orbeckst) [inline package/MDAnalysis/analysis/rms.py]: (I think you can also leave the message for `logger.exception` empty and it will show the stack trace.)  But otherwise make sure that they have the same message, e.g.
```python
errmsg = "Reference and trajectory atom ... N_ref={0:d}, N_traj={1:d}".format(...)
logger.exception(errmsg)
raise SelectionError(errmsg)
```

File: package/MDAnalysis/analysis/rms.py [hunk 6]
```diff
@@ -386,7 +386,7 @@ def __init__(self, traj, reference=None, select='all',
         for igroup, (sel, atoms) in enumerate(zip(self.groupselections,
                                                   self.groupselections_atoms)):
             if len(atoms['mobile']) != len(atoms['reference']):
-                logger.exception()
+                logger.exception('SelectionError: Group Selection')
```
Comment (orbeckst) [inline package/MDAnalysis/analysis/rms.py]: see above about logging/raising errors

File: testsuite/MDAnalysisTests/analysis/test_rms.py [hunk 7]
```diff
@@ -128,6 +129,70 @@ def test_with_superposition_equal(self):
         rmsd_superposition = rms.rmsd(A, B, center=True, superposition=True)
         assert_almost_equal(rmsd, rmsd_superposition)
 
+class TestRMSD(object):
```
Comment (orbeckst) [inline testsuite/MDAnalysisTests/analysis/test_rms.py]: Thank you!!

File: testsuite/MDAnalysisTests/analysis/test_rms.py [hunk 8]
```diff
@@ -128,6 +129,70 @@ def test_with_superposition_equal(self):
...
+        correct_values = [[  0, 0, 0],
+                          [  49, 48.9999, 4.68953]]
+        assert_almost_equal(RMSD.rmsd, correct_values, 4,
+                            err_msg="error: rmsd profile should match test " +
+                            "values")
+
+
+    def test_rmsd_single_frame(self):
+        RMSD = MDAnalysis.analysis.rms.RMSD(self.universe, select='name CA')
+        RMSD.run(start=5, stop=6)
```
Comment (orbeckst) [inline testsuite/MDAnalysisTests/analysis/test_rms.py]: `assert` something

File: testsuite/MDAnalysisTests/analysis/test_rms.py [hunk 9]
```diff
@@ -128,6 +129,70 @@ def test_with_superposition_equal(self):
...
+        self.outfile = os.path.join(self.tempdir.name, 'rmsd.npy')
+
+    def tearDown(self):
+        del self.universe
+        del self.tempdir
+
+    def test_rmsd(self):
+        RMSD = MDAnalysis.analysis.rms.RMSD(self.universe, select='name CA')
+        RMSD.run(step=49)
+        correct_values = [[  0, 0, 0],
```
Comment (orbeckst) [inline testsuite/MDAnalysisTests/analysis/test_rms.py]: store the `correct_values` in the class (a "fixture") so that you can refer to them here and in `test_mass_weighted_and_save()`.

File: testsuite/MDAnalysisTests/analysis/test_rms.py [hunk 10]
```diff
@@ -128,6 +129,70 @@ def test_with_superposition_equal(self):
...
+
+    def test_rmsd_single_frame(self):
+        RMSD = MDAnalysis.analysis.rms.RMSD(self.universe, select='name CA')
+        RMSD.run(start=5, stop=6)
+
+
+    def test_mass_weighted_and_save(self):
+        RMSD = MDAnalysis.analysis.rms.RMSD(self.universe, select='name CA')
+        RMSD.run(step=49, mass_weighted= True)
+        RMSD.save()
```
Comment (orbeckst) [inline testsuite/MDAnalysisTests/analysis/test_rms.py]: read the save file and compare to the `correct_values` (see above) to `assert` correct file writing.

File: testsuite/MDAnalysisTests/analysis/test_rms.py [hunk 11]
```diff
@@ -128,6 +129,70 @@ def test_with_superposition_equal(self):
...
+        
+    def test_rmsd_group_selections(self):
+        RMSD = MDAnalysis.analysis.rms.RMSD(self.universe,
+                                            groupselections=
+                                            ['backbone','name CA'])
+        RMSD.run(step=49)
```
Comment (orbeckst) [inline testsuite/MDAnalysisTests/analysis/test_rms.py]: `assert` something

File: testsuite/MDAnalysisTests/analysis/test_rms.py [hunk 12]
```diff
@@ -128,6 +129,70 @@ def test_with_superposition_equal(self):
...
+
+
```
Comment (orbeckst) [inline testsuite/MDAnalysisTests/analysis/test_rms.py]: excellent job testing the failure cases!

**Comment by orbeckst** [comment]
@jdetle very good job! I'd like you to attend to the comments – primarily always `assert` something.

Thanks a lot for fixing **and testing** RMSD.

File: testsuite/MDAnalysisTests/analysis/test_rms.py (hunk 10 above)
Comment (jdetle) [inline testsuite/MDAnalysisTests/analysis/test_rms.py]: Okay thanks for the review. This is stuff I can get done in some free time
tonight. I am sitting in my car about to drive up to the Bay Area as of
this very moment.
On Sat, Jul 9, 2016 at 3:16 PM Oliver Beckstein <notifications@github.com>
wrote:

> In testsuite/MDAnalysisTests/analysis/test_rms.py
> <https://github.com/MDAnalysis/mdanalysis/pull/889#discussion_r70172708>:
>
> > +        correct_values = [[  0, 0, 0],
> > +                          [  49, 48.9999, 4.68953]]
> > +        assert_almost_equal(RMSD.rmsd, correct_values, 4,
> > +                            err_msg="error: rmsd profile should match test " +
> > +                            "values")
> > +
> > +
> > +    def test_rmsd_single_frame(self):
> > +        RMSD = MDAnalysis.analysis.rms.RMSD(self.universe, select='name CA')
> > +        RMSD.run(start=5, stop=6)
> > +
> > +
> > +    def test_mass_weighted_and_save(self):
> > +        RMSD = MDAnalysis.analysis.rms.RMSD(self.universe, select='name CA')
> > +        RMSD.run(step=49, mass_weighted= True)
> > +        RMSD.save()
>
> read the save file and compare to the correct_values (see above) to assert
> correct file writing.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/MDAnalysis/mdanalysis/pull/889/files/023408db07dbe5fcb0bc71d0298fbb0b2dc11d88#r70172708>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AKcARr1QEcytOeSe-fTciWayQypUT6wPks5qUB3IgaJpZM4JGYJ2>
> .
>
-- 
Have a wonderful day,
John J. Detlefs

File: package/MDAnalysis/analysis/rms.py (hunk 4 above)
Comment (jdetle) [inline package/MDAnalysis/analysis/rms.py]: Weird...

**Comment by jdetle** [comment]
Comments attended to!

**Comment by orbeckst** [comment]
Thanks a lot @jdetle, good work!
